ports/nrf/common-hal/alarm/__init__.c

```diff
@@ -0,0 +1,236 @@
+        dbg_printf("wakeup cause = NRF_SLEEP_WAKEUP_%s\r\n",
+		   cause_str[(int)cause]);
+    }
+}
+#endif
+
+bool alarm_woken_from_sleep(void) {
+   nrf_sleep_source_t cause = _get_wakeup_cause();
+#ifdef NRF_DEBUG_PRINT
+   if (cause != NRF_SLEEP_WAKEUP_UNDEFINED) {
+       //print_wakeup_cause(cause);
+   }
+#endif
+   return (cause == NRF_SLEEP_WAKEUP_GPIO || cause == NRF_SLEEP_WAKEUP_TIMER
+	   || cause == NRF_SLEEP_WAKEUP_TOUCHPAD);
+}
+
+STATIC mp_obj_t _get_wake_alarm(size_t n_alarms, const mp_obj_t *alarms) {
+    nrf_sleep_source_t cause = _get_wakeup_cause();
+#ifdef NRF_DEBUG_PRINT
+    //print_wakeup_cause(cause);
+#endif
+    switch (cause) {
+      case NRF_SLEEP_WAKEUP_TIMER: {
+        return alarm_time_timealarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      case NRF_SLEEP_WAKEUP_TOUCHPAD: {
+        return alarm_touch_touchalarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      case NRF_SLEEP_WAKEUP_GPIO: {
+        return alarm_pin_pinalarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      default:
+        break;
+    }
+    return mp_const_none;
+}
+
+mp_obj_t common_hal_alarm_get_wake_alarm(void) {
+    mp_obj_t obj = _get_wake_alarm(0, NULL);
+    return obj;
+}
+
+// Set up light sleep or deep sleep alarms.
+STATIC void _setup_sleep_alarms(bool deep_sleep, size_t n_alarms, const mp_obj_t *alarms) {
+    alarm_pin_pinalarm_set_alarms(deep_sleep, n_alarms, alarms);
+    alarm_time_timealarm_set_alarms(deep_sleep, n_alarms, alarms);
+    alarm_touch_touchalarm_set_alarm(deep_sleep, n_alarms, alarms);
+}
+
+STATIC void _idle_until_alarm(void) {
+#ifdef NRF_DEBUG_PRINT
+    int ct = 40;
+#endif
+    reset_reason_saved = 0;
+    // Poll for alarms.
+    while (!mp_hal_is_interrupted()) {
+        RUN_BACKGROUND_TASKS;
+        // Allow ctrl-C interrupt.
+        if (alarm_woken_from_sleep()) {
+	    alarm_save_wake_alarm();
+	    return;
+	}
+        port_idle_until_interrupt();
+#ifdef NRF_DEBUG_PRINT
+        if (ct > 0) {
+            dbg_printf("_");
+            --ct;
+        }
+#endif
+    }
+}
+
+mp_obj_t common_hal_alarm_light_sleep_until_alarms(size_t n_alarms, const mp_obj_t *alarms) {
+    mp_obj_t r_obj = mp_const_none;
+    _setup_sleep_alarms(false, n_alarms, alarms);
+#ifdef NRF_DEBUG_PRINT
+    dbg_printf("\r\nsleep...");
+#endif
+
+    _idle_until_alarm();
+
+    if (mp_hal_is_interrupted()) {
+#ifdef NRF_DEBUG_PRINT
+        dbg_printf("mp_hal_is_interrupted\r\n");
+#endif
+        r_obj = mp_const_none;
+    }
+    else {
+        r_obj = _get_wake_alarm(n_alarms, alarms);
+    }
+    alarm_reset();
+    return r_obj;
+}
+
+void common_hal_alarm_set_deep_sleep_alarms(size_t n_alarms, const mp_obj_t *alarms) {
+    _setup_sleep_alarms(true, n_alarms, alarms);
+}
+
+extern void set_memory_retention(void);
+
+void NORETURN alarm_enter_deep_sleep(void) {
+    alarm_pin_pinalarm_prepare_for_deep_sleep();
+    alarm_touch_touchalarm_prepare_for_deep_sleep();
+
```

Comment:
Here, to implement a deep sleep `TimeAlarm`, don't go into System OFF, since that will shut down the RTC. Instead, stay in System ON, and set up the RTC with the prescaler to wake up in the proper amount of time. You can still use pin alarms, right? They should still work fine in System ON state, I think.